README.md
# A Node.js and browser Flickr API library

With oauth authentication for Flickr API keys if you're using it server-side (authenticated calls from the browser are too insecure to support for the moment, and will throw an error).

You also get API route proxying so you can call the Flickr methods through your own server and get Flickr responses back for free. Super handy!

# Quick start guide:

You don't want to read this entire README.md, so just find what you want to do in this handy quick start guide right here at the start of the README, and off you go!

## In the browser with client-side credentials:

Script-load the `browser/flickrapi.dev.js` library for development work, and use the `browser/flickrapi.js` library in production.

You can access Flickr by creating an API instance as:

```
var flickr = new Flickr({
  api_key: "1234ABCD1234ABCD1234ABCD1234ABCD"
});
```

Then query Flickr using the API as described over at http://www.flickr.com/services/api - for instance, to search for all photographs that text-match the terms "red panda", you call:

```
flickr.photos.search({
  text: "red+panda"
}, function(err, result) {
  if(err) { throw new Error(err); }
  // do something with result
}
```
All calls are asynchronous, and the callback handling function always has two arguments. The first, if an error occurs, is the error generated by Flickr; the second, if the call succeeds, is the result sent back from Flickr, as plain JavaScript object.

**Note**: this is not secure. People will be able to see your API key, and this is pretty much *the worst idea(tm)*, so you probably want to use this library...

## In the browser, securely via proxied API calls:

Script-load the `browser/flickrapi.dev.js` library for development work, and use the `browser/flickrapi.js` library in production, but don't use your API key. Instead, point to your server as a flickr API proxy:

```
var flickr = new Flickr({
  endpoint: "http//yourhostedplace/services/rest/"
});
```

To make this work, have flickapi running on your server with a proxy route enabled, and you'll be able to make use of all the Flickr API calls, without having to put your credentials anywhere in your client-side source code.

Proxy mode is explained [below](#Flickr_API_proxying_for_connect/express_apps), but is essentially a one-liner add to your regular connect/express app.

## As a module in a (larger) Node.js program

Install like any other package:
```
$> npm install flickrapi --save
```
After that, you have two choices, based on whether you want to authenticate or not. Both approaches require an API key, but using OAuth2 authentication means you get access to the full API, rather than only the public API.

To suppress the progress bars in stdout you can include a `progress` attribute when initializing:

```
var flickr = new Flickr({
    api_key: "1234ABCD1234ABCD1234ABCD1234ABCD",
    progress: false
});
```

### No authentication, public API only

```
var Flickr = require("flickrapi"),
    flickrOptions = {
      api_key: "API key that you get from Flickr",
      secret: "API key secret that you get from Flickr"
    };

Flickr.tokenOnly(flickrOptions, function(error, flickr) {
  // we can now use "flickr" as our API object,
  // but we can only call public methods and access public data
});
```

### Authenticated, full Flickr API access

```
var Flickr = require("flickrapi"),
    flickrOptions = {
      api_key: "API key that you get from Flickr",
      secret: "API key secret that you get from Flickr"
    };

Flickr.authenticate(flickrOptions, function(error, flickr) {
  // we can now use "flickr" as our API object
});
```

# The end.

That's it, that's all the quickstart guiding you need. For more detailed information, keep reading. If you just wanted to get up and running, then the preceding text should have gotten you there!

---

# What to do now that you have the library loaded


## Call some API functions

calling API functions is then a matter of calling the functions as they are listed on http://www.flickr.com/services/api, so if you wish to get all your own photos, you would call:

```
flickr.photos.search({
  user_id: flickr.options.user_id,
  page: 1,
  per_page: 500
}, function(err, result) {
  // result is Flickr's response
});
```
## Call functions that don't *require* authentication, authenticated anyway

Simply add an `authenticated: true` pair to your function call. Compare:

```
flickr.people.getPhotos({
  api_key: ...
  user_id: <your own ID>
  page: 1,
  per_page: 100
}, function(err, result) {
  /*
    This will give public results only, even if we used
    Flickr.authenticate(), because the function does not
    *require* authentication to run. It just runs with
    fewer permissions.
  */
});
```

To:

```
flickr.people.getPhotos({
  api_key: ...
  user_id: <your own ID>
  authenticated: true,
  page: 1,
  per_page: 100
}, function(err, result) {
  /*
    This will now give all public and private results,
    because we explicitly ran this as an authenticated call
  */ 
});
```


## Proxy the Flickr API

If your app is a [connect](https://github.com/senchalabs/connect) or [express](http://expressjs.com) app, you get Flickr API proxying
for free.

Simply use the `.proxy()` function to set everything up and then call your own API route in the same way you would call the Flickr API, minus the security credentials, since the servers side Flickr api object already has those baked in.

As an example, the test.js script for node-flickrapi uses the following code to set up the local API route:

```
var express = require("express");

Flickr.authenticate(FlickrOptions, function(error, flickr) {
  var app = express();
  app.configure(function() {
    ...
    flickr.proxy(app, "/service/rest");
    ...
  });
  ...
});
```

This turns the `/service/rest` route into a full Flickr API proxy, which the browser library can talk to, using `POST` operations.

To verify your proxy route works, simply use `cURL` in the following fashion:

```
curl -X POST -H "Content-Type: application/json"
             -d '{"method":"flickr.photos.search", "text":"red+pandas"}'
             http://127.0.0.1:3000/service/rest/
```

Note that the proxy is "open" in that there is no explicit user management. If you want to make sure only "logged in users" get to use your API proxy route, you can pass an authentication middleware function as third argument to the `.proxy` function:

```
function authenticator(req, res, next) {
  // assuming your session management uses req.session:
  if(req.session.authenticated) {
    return next();
  }
  next({status:403, message: "not authorised to call API methods"});
}

flickr.proxy(app, "/service/rest/", authenticator);
```


## Upload photos to Flickr

If you're running the code server-side, and you've authenticated with Flickr already, you can use the `Flickr.upload` function to upload individual photos, or batches of photos, to your own account (or, the account that is tied to the API key that you're using).

```
Flickr.authenticate(FlickrOptions, function(error, flickr) {
  var uploadOptions = {
    photos: [{
      title: "test",
      tags: [
        "happy fox",
        "test 1"
      ],
      photo: __dirname + "/test.jpg"
    },{
      title: "test2",
      tags: "happy fox image \"test 2\" separate tags",
      photo: __dirname + "/test.jpg"
    }]
  };

  Flickr.upload(uploadOptions, FlickrOptions, function(err, result) {
    if(err) {
      return console.error(error);
    }
    console.log("photos uploaded", result);
  });
});
```

For the list of available upload properties, see the [Flickr Upload API](https://www.flickr.com/services/api/upload.api.html) page.

## Download data from Flickr

You can use this module to very easily download all your Flickr content, using the built in `downsync` function:

```
var Flickr = require("flickrapi"),
    flickrOptions = { ... };
Flickr.authenticate(flickrOptions, flickrapi.downsync());
```

That's all you need to run. The package will generate a data directory with your images in `./data/images` (in several sizes), and the information architecture (metadata, sets, collections, etc) in `./data/ia`.

If you want this in a different directory, you can pass the dir as an argument to the downsync function:

```
var Flickr = require("flickrapi"),
    flickrOptions = { ... };
Flickr.authenticate(flickrOptions, flickrapi.downsync("userdata/me"));
```

This will now create a `./data` for the flickr API information, but also a `./userdata/me/` directory that contains the `images` and `ia` dirs with your personal data.

### Downloading shortcut: one-step downsyncing

If you just want to immediately downsync all your data right now, simply use the `test.js` application with the --downsync runtime argument: add your Flickr API key information to the `.env` file and then run:

`$> node test --downsync`

Run through the authentication procedure, and then just wait for it to finish. Once it's done, you should have a local mirror of all your Flickr data.

#### (Re)syncing with Flickr in your code

(Re)syncing is a mostly a matter or running the downsync function again. This will update anything that was updated or added on Flickr, but will not delete anything from your local mirror that was deleted from Flickr unless specifically told to do so, by passing a second argument (internally known as the "removeDeleted" flag in the code) to the `downsync` function call:

```
var Flickr = require("flickrapi"),
    flickrOptions = { ... };
Flickr.authenticate(flickrOptions, flickrapi.downsync("userdata/me", true));
```

If `true`, this will delete local files that were removed on Flickr (e.g. photos that you didn't like anymore, etc). If `false`, or omitted, no pruning of the local mirror will be performed.

## Use your Flickr data in another application

If you downloaded all your Flickr data, you can use these in your own node apps by "dry loading" Flickr:

```
var Flickr = require("flickrapi"),
    flickrData = Flickr.loadLocally();
```

This will give you an object with the following structure:

```
{
  photos: [photo objects],
  photo_keys: [photo.id array, sorted on publish date],
  photosets: [set objects],
  photoset_keys: [set.id array, sorted on creation date],
  collections: [collection objects],
  collection_keys: [collection.id array, sorted on title],
}
```

Not sure what these objects look like? head over to your `./data/ia` directory and just open a .json file in your favourite text editor.

The `loadLocally` function can take two arguments, namely a location where the ia data can be found, and an options object. If you want to pass in an options object you *must* supply a location, too.

```
flickrData = Flickr.loadLocally("./userdata", {
  loadPrivate: false
});
```

Currently the options object only has one meaningful property, `loadPrivate`, which determines whether or not photos and photosets that are marked "not public" in Flickr show up in the `photo_keys` and `photoset_keys` lists.


# An example of a first run

## Fetching Flickr's most up to date API definitions

On first run, the package will fetch all known methods from Flickr, and cache them for future use. This can take a bit, as there are a fair number of methods, but is inconsequential on subsequent package loading.

## Authenticate your API key with Flickr

On first run, the authentication function will notice that there are no `access_token` and `access_token_secret` values set, and will negotiate these with Flickr using their oauth API, based on the permissions you request for your API key.

By default, the only permissions are "read" permissions, but you can override this by adding a `permissions` property to the options object:

* `permissions: "read"` will give the app read-only access (default)
* `permissions: "write"` will give it read + write access
* `permissions: "delete"` will give it read, write and delete access

Note that you cannot make use of the upload functions unless you authenticate with `write` or `delete` permissions.

### An example run

Running the app will show output such as the following block:

```
$> node app
{ oauth_callback_confirmed: 'true',
  oauth_token: '...',
  oauth_token_secret: '...' }
prompt: oauth_verifier: _
```

Once the app reaches this point it will open a browser, allowing you to consent to the app accessing your most private of private parts. On Flickr, at least. If you agree to authorize it, you will get an authorisation code that you need to pass so that the flickrapi can negotiate access tokens with Flickr.

Doing so continues the program:

```
$> node app
{ oauth_callback_confirmed: 'true',
  oauth_token: '...',
  oauth_token_secret: '...' }
prompt: oauth_verifier: 123-456-789

Add the following variables to your environment:

export FLICKR_USER_ID="12345678%40N12"
export FLICKR_ACCESS_TOKEN="72157634942121673-3e02b190b9720d7d"
export FLICKR_ACCESS_TOKEN_SECRET="99c038c9fc77673e"
```

These are namespaced environment variables, which works really well with env packages like [habitat(https://www.npmjs.com/package/habitat),
so if you're going to use a namespace-aware enviroment parser, simply add these variables
to your environment, or put them in an `.env` file and then parse them in.

If you would prefer to use plain `process.env` consulting, remove the `FLICKR_` namespace
prefix, and then pass `process.env` as options object.

Alternatively, if you don't mind hardcoding values (but **be careful never to check that code
in, because github gets mined by bots for credentials**) you can put them straight into your
source code:

```
var FlickrOptions = {
      api_key: "your API key",
      secret: "your API key secret",
      user_id: "...",
      access_token: "...",
      access_token_secret: "..."
    }
```

The flickrapi package will now be able to authenticate with Flickr without constantly needing to ask you for permission to access data.

## Using your own OAuth callback endpoint

By default the oauth callback is set to "out-of-band". You can see this in the `.env` file as the `FLICK_CALLBACK="oob"` parameter, but if this is omitted the code falls back to oob automatically. For automated processes, or if you don't want your uers to have to type anything in a console, you can override this by setting your own oauth callback endpoint URL.

Using a custom callback endpoint, the oauth procedure will contact the indicated endpoint with the authentication information, rather than requiring your users to manually copy/paste the authentication values.

**Note** your users will still need to authenticate the app from a browser!

To use a custom endpoint, add the URL to the options as the `callback` property:

```
var options = ...;
options.callback: "http://.../...";
Flickr.authenticate(options, function(error, flickr) {
  ...
}
```

You can make your life easier by using an environment variable in the `.env` file rather than hardcoding your endpoint url:

```
export FLICKR_CALLBACK="http://..."
```

The callback URL handler will at its minimum need to implement the following middleware function:

```
function(req, res) {
  res.write("");
  options.exchange(req.query);
}
```

However, having the response tell the user that authorisation was received and that they can safely close this window/tab is generally a good idea.

If you wish to call the exchange function manually, the object expected by `options.exchange` looks like this:

```
{
  oauth_token: "...",
  oauth_verifier: "..."
}
```

# Advanced topics


CHANGELOG
v0.4.1
  * URL data now conforms to RFC5987 (titles, descriptions, etc)

v0.3.35
  * tags can now include the apostrophe (') character

v0.3.33
  * Authenticate is now exiting, rather than wrapping the entire
    life of the API object.

v0.3.31
  * Uploads can use tags either in string form or in array form.

v0.3.30
  * API functions that have optional authentication (such as
    flickr.people.getPhotos) can now be told to explicitly use
    authenticated calls, by passing in "authenticated: true"
    as part of the function call options.

v0.3.24
  * The library can now be used to upload photos to Flickr.

v0.3.23
  * private photos are now downloaded during --downsync, provided
    that the `force_auth` option is set in the flickr options, with
    the ia function (src/handlers/ia.js) now having an options
    object that can take a "loadPrivate" boolean property. If
    false or omitted, private photos and photosets are filtered
    out of the keysets for each. If true, photos and photosets
    are left unfiltered and will show up if you do things based
    on photo_keys and photoset_keys. (see README.md)

v0.3.22
  * Rewrote the downsync code to sommething a little more
    sensible, and changed the downsync order from
    photos/sets/collections to collections/sets/photos,
    so that changes to the collection and set metadata aren't
    held up by potentially thousands of photograph checks.

v0.3.21
  * Added "force_auth" as possible flickrOptions property,
    which allows you to force authed API calls for functions
    that do not require auth, but may generate more data if
    authed (such as the photos.search function, which does
    public searching without auth, but can return private
    data when authed).

v0.3.20
  * fixed some remaining http: protocol URLs to be https:

v0.3.19
  * fixed lingering bug in tokenOnly() behaviour

v0.3.18
  * switch image download protocol from http: to https: due
    to a change in Flickr's infrastructure to https-only.

v0.3.17
  * added catching "no response" errors when querying Flickr.

v0.3.14
  * Added "token only" access, so that the public Flickr API
    calls do not require you to authenticate through oauth.

v0.3.13
  * Updated the authentication endpoints to the correct endpoints

v0.3.12
  * fixed infinite retries on downloading images

v0.3.11
  * Added additional options to control the authentication
    procedure in terms of console logging, opening browsers,
    etc. See the Custom authentication section in the README
    for additional details.

v0.3.8
  * Image downloads from Flickr that result in incomplete images
    will make the download module redownload the file. For JPG
    images this means looking for 0xFFD9 as last two bytes,
    for PNG files the code looks for the IEND chunk in the PNG
    data. No more 0 byte or half-finished images (hopefully).

v0.3.0
  * In order to align the module with the Flickr API's property
    names, the "key" property has been renamed to "api_key" in
    the client-side library, as well as the server-side module.
    As this breaks upgrades, the version has been moved up from
    0.2 to 0.3
  * The clientside library can be given a custom endpoint to
    make it query proxies for the Flickr API, to get around
    the security problems that stem from having a client side
    library needing to use private credentials.

v0.2.27
  * oauth authentication now allows you to specify your own
    callback endpoint. If not specified, the out-of-band method
    is still used, requiring users to type an authentication
    code that they receive in the browser. Otherwise all they
    have to do is press the "I authorise this app" button.
  * updated the test.js file to allow one-step downsyncing,
    to save people the hassle of having to write an app around
    the module. If you just want your flickr data, add your
    Flickr API key information to the .env file, and then call
    "node test --downsync"

v0.2.26
  * Flickr API proxying, so that your app can use query flickr
    server side and send the response back over your own API
    url. Because client-side Flickr-API calls cannot be signed
    securely.

v0.2.25
  * client side improvment: much smaller production library,
    acceptably sized dev library.
  * README documentation ammended for client-side library use.

v0.2.22
  * also compiles to client-side library now, with a precompiled
    library in the ./browser directory.

v0.2.21
  * all code uses strict pragma, and passes linting.

v0.2.18
  * added collection icon image mirroring, so you don't need
    to create your own pinhead collages.

v0.2.16
  * "on downsync complete" callback added

v0.2.14
  * downsync will now prune entries that you removed from Flickr,
    but still have in your mirror, but only when the `downsync`
    function is called with the option second `removeDeleted`
    parameter set to `true`.

v0.2.11
  * updated the progress package. It's now well-behaved on OSX.

v0.2.9
  * downsync now updates entries on disk that were updated on Flickr
    since the last downsync, rather than requiring you delete your
    ia directory and redownloading all the .json data.

v0.2.8
  * photos are correctly sorted in photosets

v0.2.6
  * comments and contexts are cross-referenced into photos

v0.2.5
  * functions calls are always correctly auth'd
  * app permissions can be passed during the `authenticate` call,
    defaulting to "read".

v0.2.1
  * downsync now downloads:
    - comments for each photo
    - contexts for each photo
  * "large" and "medium 800" images are downloaded for each photo
  * photos now indicate the image sizes that Flickr has for them

v0.1.x
  * downsync downloads:
    - photos
    - photosets
    - collections

Versions not listed are bug fix releases, rather than feature releases.


src/utils.js
/**
 *  This is a utility object for universal oauth
 *  signing as well as querying Flickr once a query
 *  object has been constructed to set to the Flickr
 *  API endpoint.
 *
 *  Response are in JSON format.
 */
module.exports = (function() {
  "use strict";
  var crypto = require("crypto"),
      fs = require("fs"),
      request = require("request"),
      statusCodes = {
        400: "Bad Request",
        401: "Unauthorized",
        402: "Payment Required",
        403: "Forbidden",
        404: "Not Found",
        405: "Method Not Allowed",
        406: "Not Acceptable",
        407: "Proxy Authentication Required",
        408: "Request Timeout",
        409: "Conflict",
        410: "Gone",
        411: "Length Required",
        412: "Precondition Failed",
        413: "Request Entity Too Large",
        414: "Request-URI Too Long",
        415: "Unsupported Media Type",
        416: "Requested Range Not Satisfiable",
        417: "Expectation Failed",
        428: "Precondition Required",
        429: "Too Many Requests",
        431: "Request Header Fields Too Large",
        450: "Blocked By Windows Parental Controls",
        499: "Client Closed Request",
        500: "Internal Server Error",
        501: "Not Implemented",
        502: "Bad Gateway",
        503: "Service Unavailable",
        504: "Gateway Timeout",
        505: "HTTP Version Not Supported",
        506: "Variant Also Negotiates",
        507: "Insufficient Storage",
        508: "Loop Detected",
        509: "Bandwidth Limit Exceeded",
        510: "Not Extended",
        511: "Network Authentication Required"
      };

  /**
   * Pretty-print JSON files, because we will want
   * to inspect them manually, as good humans.
   */
  module.exports = (function() {
    if (!JSON.prettyprint) {
      JSON.prettyprint = function prettyprint(data) {
        return this.stringify(data, undefined, 2);
      };
    }
    return JSON;
  }());

  var callErrors = {};

  return {
    /**
     * shorthand function
     */
    mkdir: function(dir) {
      var trymkdir = function(dir) {
        try {
          fs.mkdirSync(dir);
          //console.log("creating " + dir);
        }
        catch (e) {
          /* we really don't care if it already exists */
        }
      };
      var f = "";
      dir.replace("./",'').split("/").forEach(function(d) {
        f += d + "/";
        trymkdir(f);
      });
      return dir;
    },

    /**
     * extend the known generic Flickr errors
     */
    extendErrors: function(errors, errCap) {
      errors.forEach(function(err) {
        if (+(err.code) >= errCap) {
          callErrors[err.code] = err;
        }
      });
    },

    getCallErrors: function() {
      return callErrors;
    },

    /**
     * Update an options object for use with Flickr oauth
     * so that it has a new timestampe and nonce.
     */
    setAuthVals: function(options) {
      var timestamp = "" + Date.now(),
          md5 = crypto.createHash('md5').update(timestamp).digest("hex"),
          nonce = md5.substring(0,32);
      options.oauth_timestamp = timestamp;
      options.oauth_nonce = nonce;
      return options;
    },

   /**
    * Properly encode query strings with RFC5987 values.
    * Additionally, RFC3986 and Flickr require the exclamation point be escaped.
    * Inspired by MDN: https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent
    */
   encodeRFC5987ValueChars: function(str) {
     return encodeURIComponent(str).
     replace(/['()!]/g, escape).
     replace(/\*/g, '%2A').
     replace(/%(?:7C|60|5E)/g, unescape);
   },

   /**
    * Collapse a number of oauth query arguments into an
    * alphabetically sorted, URI-safe concatenated string.
    */
   formQueryString: function(queryArguments) {
     var Utils = this,
         args = [],
         append = function(key) {
           args.push(key + "=" + Utils.encodeRFC5987ValueChars(queryArguments[key]));
         };
     Object.keys(queryArguments).sort().forEach(append);
     return args.join("&");
   },

    /**
     * Turn a url + query string into a Flickr API "base string".
     */
    formBaseString: function(verb, url, queryString) {
      return [verb, encodeURIComponent(url), encodeURIComponent(queryString)].join("&");
    },

    /**
     * Parse a Flickr API response.
     */
    parseRestResponse: function(body) {
      if(!body) {
        return false;
      }
      var constituents = body.split("&"),
          response = {},
          keyval;
      constituents.forEach(function(pair) {
        keyval = pair.split("=");
        response[keyval[0]] = decodeURIComponent(keyval[1]);
      });
      return response;
    },

    /**
     * HMAC-SHA1 data signing
     */
    sign: function(data, key, secret) {
      var hmacKey = key + "&" + (secret ? secret : ''),
          hmac = crypto.createHmac("SHA1", hmacKey);
      hmac.update(data);
      var digest = hmac.digest("base64");
      return encodeURIComponent(digest);
    },

    /**
     * Validate an api call
     */
    checkRequirements: function(method_name, required, callOptions, callback) {
      for(var r=0, last=required.length, arg; r<last; r++) {
        arg = required[r];
        if(arg.name === "api_key") continue;
        if(!callOptions.hasOwnProperty(arg.name)) {
          return callback(new Error("missing required argument '"+arg.name+"' in call to "+method_name));
        }
      }
    },

    /**
     * Generate the query arguments for querying flickr
     */
    generateQueryArguments: function(method_name, flickrOptions, callOptions) {
      // set up authorized method access
      var queryArguments = {
        method: method_name,
        api_key: flickrOptions.api_key
      };

      // set up bindings for method-specific args
      if (callOptions) {
        Object.keys(callOptions).forEach(function(key) {
          queryArguments[key] = callOptions[key];
        });
      }

      return queryArguments;
    },

    /**
     * Generate an API function
     */
    generateAPIFunction: function(method_name, flickrOptions, security, required, optional, errors) {
      var Utils = this, fn;

      // Will we need to be authenticated to call this function?
      if(flickrOptions.tokenonly && security.needslogin===1) {
        fn = function(callOptions, callback) {
          if(callOptions && !callback) { callback = callOptions; callOptions = {}; }
          callback("the "+method_name+" function requires authentication, and can only be called through an authenticated Flickr API instance.");
        };
      }

      // code path for compiling the client-side JS library
      else if(typeof process.CLIENT_COMPILE !== "undefined") {
       fn = function(callOptions, callback) {
          if(callOptions && !callback) { callback = callOptions; callOptions = {}; }
          var queryArguments = Utils.generateQueryArguments(method_name, this.flickrOptions, callOptions);
          Utils.queryFlickr(queryArguments, this.flickrOptions, security, callback);
        };
      }

      // server-side code path
      else {
        fn = function(callOptions, callback) {
          if(callOptions && !callback) { callback = callOptions; callOptions = {}; }
          Utils.checkRequirements(method_name, required, callOptions, callback);
          var queryArguments = Utils.generateQueryArguments(method_name, flickrOptions, callOptions);
          Utils.queryFlickr(queryArguments, flickrOptions, security, callback, errors);
        };
      }

      fn.data = {
        required: required,
        optional: optional,
        errors: errors,
        security: security,
        name: method_name,
        url: "https://www.flickr.com/services/api/"+method_name+".html"
      };

      return fn;
    },

    /**
     * Call the Flickr API
     */
    queryFlickr: function(queryArguments, flickrOptions, security, processResult, errors) {
      queryArguments = JSON.parse(JSON.stringify(queryArguments));

      if(arguments.length === 3) {
        processResult = arguments[2];
        security = {};
        errors = {};
      }

      var authed = (security.needssigning === 1) || queryArguments.authenticated || flickrOptions.force_auth;

      // do we need to HMAC-SHA1 sign this URL?
      if (authed) {
        flickrOptions = this.setAuthVals(flickrOptions);
        queryArguments.oauth_nonce = flickrOptions.oauth_nonce;
        queryArguments.oauth_timestamp = flickrOptions.oauth_timestamp;
        queryArguments.oauth_consumer_key = flickrOptions.api_key;
        queryArguments.oauth_token = flickrOptions.access_token;
        queryArguments.oauth_signature_method = "HMAC-SHA1";
      }

      // nope - plain API key will suffice
      else {
        queryArguments.api_key = flickrOptions.api_key;
      }

      // force JSON request
      queryArguments.format = "json";

      var url = "https://api.flickr.com/services/rest/",
          queryString = this.formQueryString(queryArguments),
          data = this.formBaseString("GET", url, queryString),
          signature = authed ? "&oauth_signature=" + this.sign(data, flickrOptions.secret, flickrOptions.access_token_secret) : '',
          flickrURL = url + "?" + queryString + signature;

      request.get(flickrURL, function(error, response, body) {

        if(!response) {
          error = "HTTP Error: no response for url [" + flickrURL + "]";
          if (flickrOptions.retry_queries) {
            return this.queryFlickr(queryArguments, flickrOptions, security, processResult, errors);
          }
          return processResult(error);
        }

        if(!body) {
          error = "HTTP Error " + response.statusCode + " (" + statusCodes[response.statusCode] + ")";
          return processResult(error);
        }

        // we can transform the error into something more
        // indicative if "errors" is an array of known errors
        // for this specific method call.
        if(!error) {
          try {
            body = body.trim().replace(/^jsonFlickrApi\(/,'').replace(/\}\)$/,'}');
            body = JSON.parse(body);
            if(body.stat !== "ok") {
              return processResult(new Error(body.message));
            }
          } catch (e) {
            return processResult("could not parse body as JSON: " + body);
          }
        }

        processResult(false, body);
      }.bind(this));
    },

    /**
     * Call the Flickr API for uploading a photo. Uploading is always authenticated.
     *
     *  photo: The file to upload.
     *
     *  title (optional): The title of the photo.
     *  description (optional): A description of the photo. May contain some limited HTML.
     *  tags (optional): A space-seperated list of tags to apply to the photo.
     *  is_public, is_friend, is_family (optional): Set to 0 for no, 1 for yes. Specifies who can view the photo.
     *  safety_level (optional): Set to 1 for Safe, 2 for Moderate, or 3 for Restricted.
     *  content_type (optional): Set to 1 for Photo, 2 for Screenshot, or 3 for Other.
     *  hidden (optional): Set to 1 to keep the photo in global search results, 2 to hide from public searches.
     *
     */
    upload: function(uploadOptions, flickrOptions, processResult) {
      // upload oauth uses "everything except the photo itself"
      if(!uploadOptions.photos && !uploadOptions.photo) {
        return processResult("upload requires at least one photo is passed for uploading.");
      }
      if(!uploadOptions.photos) {
        uploadOptions.photos = [ uploadOptions.photo ];
      }
      var photos = uploadOptions.photos;
      var uploadids = [];
      delete uploadOptions.photos;
      var self = this;

      (function next(err, result) {
        if(err) { return processResult(err); }
        if(result) { uploadids.push(result); }
        if(photos.length === 0) { return processResult(false, uploadids); }
        self.uploadtoFlickr(photos.splice(0,1)[0], flickrOptions, next);
      }(false, false));

    },

    uploadtoFlickr: function(photoOptions, flickrOptions, callback) {
      var photo = photoOptions.photo;
      delete photoOptions.photo;
      if(typeof photo === "string") { photo = fs.createReadStream(photo); }

      // collapse tags, if used
      if(photoOptions.tags && photoOptions.tags.forEach) {
        photoOptions.tags = photoOptions.tags.map(function(v) {
          return '"' + v.replace(/'/g,'%27') + '"';
        }).join(" ");
      }

      // perform title replacement for quotes
      if(photoOptions.title) {
        photoOptions.title = photoOptions.title.replace(/'/g,'%27').replace(/"/g,'%22')
      }

      flickrOptions = this.setAuthVals(flickrOptions);
      photoOptions.oauth_signature_method = "HMAC-SHA1";
      photoOptions.oauth_consumer_key = flickrOptions.api_key;
      photoOptions.oauth_token = flickrOptions.access_token;
      photoOptions.oauth_nonce = flickrOptions.oauth_nonce;
      photoOptions.oauth_timestamp = flickrOptions.oauth_timestamp;

      // craft the authentication signature
      var url = "https://up.flickr.com/services/upload/";
      var queryString = this.formQueryString(photoOptions);
      var data = this.formBaseString("POST", url, queryString);
      photoOptions.oauth_signature = this.sign(data, flickrOptions.secret, flickrOptions.access_token_secret);

      // now we can put the photo back in
      photoOptions.photo = photo;

      // restore the percentage encoded quotes in tags and titles
      photoOptions.tags = photoOptions.tags.replace(/%27/g,"'");
      photoOptions.title = photoOptions.title.replace(/%27/g,"'").replace(/%22/g,'"');

      // and finally, form the URL we need to POST to
      var signature = "&oauth_signature=" + photoOptions.oauth_signature;
      var flickrURL = url + "?" + queryString + signature;

      var req = request.post(flickrURL, function(error, response, body) {
        // format:json does not actually work, so we need to grab the photo ID from the response XML:
        // <?xml version="1.0" encoding="utf-8" ?>\n<rsp stat="ok">\n<photoid>.........</photoid>\n</rsp>\n
        var data;
        if(!body) {
          error = error || "No body found in response";
        } else if (body.indexOf('rsp stat="ok"')>-1) {
          data = parseInt(body.split("<photoid>")[1].split("</photoid>")[0], 10);
        }
        callback(error, data);
      });
      var form = req.form();
      Object.keys(photoOptions).forEach(function(prop) {
        form.append(prop, photoOptions[prop]);
      });
    }
  };
}());
